lib/agent/redirect.js

```diff
@@ -0,0 +1,140 @@
+'use strict'
+
+const { InvalidArgumentError } = require('../core/errors')
+const util = require('../core/util')
+const assert = require('assert')
+
+const kAgent = Symbol('agent')
+
+class RedirectHandler {
+  constructor (agent, opts, handler) {
+    this.agent = agent
+    this.location = null
+    this.abort = null
+    this.opts = opts
+    this.handler = handler
+    this.destroyed = false
+  }
+
+  destroy (err) {
+    this.destroyed = true
+
+    if (this.abort) {
+      this.abort()
+    }
+
+    this.handler.onError(err)
+  }
+
+  get (origin) {
+    return this[kAgent].get(origin)
+  }
+
+  onConnect (abort) {
+    if (this.destroyed) {
+      abort()
+    } else {
+      this.abort = abort
+      this.handler.onConnect(abort)
+    }
+  }
+
+  onUpgrade (statusCode, headers, socket) {
+    assert(!this.location, 'redirect should not upgrade')
+
+    this.handler.onUpgrade(statusCode, headers, socket)
+  }
+
+  onError (error) {
+    this.handler.onError(error)
+  }
+
+  onHeaders (statusCode, headers, resume) {
+    if ([300, 301, 302, 303, 307, 308].indexOf(statusCode) === -1) {
+      return this.handler.onHeaders(statusCode, headers, resume)
+    }
+
+    this.location = parseLocation(headers)
+
+    if (!this.location) {
+      return this.handler.onHeaders(statusCode, headers, resume)
+    }
+
+    const { origin, pathname, search } = util.parseURL(new URL(this.location, this.opts.origin))
+    const path = search ? `${pathname || '/'}${search || ''}` : pathname
+
+    this.opts = { ...this.opts }
+
+    this.opts.maxRedirections = this.opts.maxRedirections - 1
+    // Remove headers referring to the original URL.
+    // By default it is Host only, unless it's a 303 (see below), which removes also all Content-* headers.
+    // https://tools.ietf.org/html/rfc7231#section-6.4
+    this.opts.headers = cleanRequestHeaders(this.opts.headers, statusCode === 303)
+    this.opts.path = path
+    this.opts.origin = origin
+
+    // https://tools.ietf.org/html/rfc7231#section-6.4.4
+    // In case of HTTP 303, always replace method to be either HEAD or GET
+    if (statusCode === 303 && this.opts.method !== 'HEAD') {
+      this.opts.method = 'GET'
+      this.opts.body = null
+    }
+  }
+
+  onData (chunk) {
+    if (this.location) {
+      // Ignore body.
+      // TODO (fix): Spec ref?
```

Comment:
@ShogunPanda do we have a spec ref in regards to what to do with body?

Reply:
So, I've read the spec again. RFC 7231, Section 6.4 
It says that redirection is used to serve the requested resource from another URL, I think it assumes that no body is generated (and thus can be ignored). Even thought generating a body is not prohibited. There is an exception to this, the status 300.

For status 301, 302, 303, 307 and 308 (the latter from RFC 7238), the spec(s) mentioned that the body usually (which means it's optional and not mandated) contain just an hyperlink to the value of the Location response header, so probably we can just ignore it.

For status 300, which is "Multiple Choices", the spec mentions both generating a Location response header AND a response body with the other possible location to follow. The interesting thing is that the spec explicitily chooses not to specify a format for such body and leave it to servers and browsers implementors. Which means, in our case, that probably we can just ignore it. (we could, in theory, let the user specify a closure and let them deal with parsing, but I think it's overly complicated).

Reply:
Can we add some notes in the code on what we do and why?

Reply:
Done in #621